doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+# ADR007: Flattening
+
+| Revision | Date       | Author           |
+| :------- | :--------- | :--------------- |
+| 1        | 2023-08-01 | Gabriela Moreira |
+
+## 1. Summary
+
+This document describes a new strategy for flattening modules, that is, replacing imports, instances and exports with the actual definitions. This is a necessary step for compilation and integration with apalache, since both the compiler and apalache cannnot handle imports, instances and exports.
+
+## 2. Context
+
+Our current strategy for flattening has intrinsic problems, and some issues that popped up cannot really be solved with the current approach. The first of those issues, and perhaps the one that better flags the problem in the current implementation, is [Flatenning breaks when importing by name a definition that depends on others #836](https://github.com/informalsystems/quint/issues/836).
+
+The existing flattener takes import/instance/export statements and replaces them with a copy of the definitions from that module, assigning new ids to those definitions and its inner IR components and adding namespaces when necessary. With this, there are two possible scenarios:
+
+1. All imports, instances and exports include all definitions (i.e. there is no `import A.foo`). In this case, we will end up with a module with all of the definitions from all used modules.
+2. There is some statement refering to a specific definition like `import A.foo`. In this case, the other definitions from `A` won't be copied, and if `foo` depends on something else, the flattened module won't be valid.
```

Comment:
Could we solve this by doing inlining on expressions, rather than manipulating modules namespaces? E.g.,

```
module Foo {
   pure val e = 1
   def foo() = e + 1
}

module Bar {
   import Foo.foo
   def bar() = foo()
}
```

Then, while flattening, when we encounter the need to import `Foo.foo` could we just inline foo first to `foo() = 1 + 1`, then add this inlined `foo` into the importing bar?

```
modules Bar {
  def foo() = 1 + H
  def bar() = foo()
}
```

Reply:
Well, for the specific case of importing one definition by name, I think we could. But for the general case, I don't think we want to do inlining on expressions, and the general solution for manipulating modules namespaces also works for this.

I don't know why exactly, but I've been trying to stay away from inlining on expressions when working on this. I think @konnov said something about not doing inlining on expressions, but I don't remember if it was something like "we shouldn't do it" or "we don't need it".